test/cql-pytest/test_group_by.py

```diff
@@ -0,0 +1,152 @@
+#############################################################################
+
+import pytest
+from util import new_test_table
+from cassandra.protocol import InvalidRequest
+
+# table1 has some pre-set data which the tests below SELECT on (the tests
+# shouldn't write to it).
+@pytest.fixture(scope="module")
+def table1(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "p int, c1 int, c2 int, v int, PRIMARY KEY (p, c1, c2)") as table:
+        stmt = cql.prepare(f'INSERT INTO {table} (p, c1, c2, v) VALUES (?, ?, ?, ?)')
+        cql.execute(stmt, [0, 0, 0, 1])
+        cql.execute(stmt, [0, 0, 1, 2])
+        cql.execute(stmt, [0, 1, 0, 3])
+        cql.execute(stmt, [0, 1, 1, 4])
+        cql.execute(stmt, [1, 0, 0, 5])
+        cql.execute(stmt, [1, 0, 1, 6])
+        cql.execute(stmt, [1, 1, 0, 7])
+        cql.execute(stmt, [1, 1, 1, 8])
+        # FIXME: I would like to make the table read-only to prevent tests
+        # from messing with it accidentally.
+        yield table
+
+### GROUP BY without aggregation:
+
+# GROUP BY partition key picks a single row from each partition, the
+# first row (in the clustering order).
+def test_group_by_partition(cql, table1):
+    assert {(0,0,0,1), (1,0,0,5)} == set(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p'))
+
+# Similarly, GROUP BY a partition key plus just a prefix of the clustering
+# key retrieves the first row in each of the clustering ranges
+def test_group_by_clustering_prefix(cql, table1):
+    assert {(0,0,0,1), (0,1,0,3), (1,0,0,5), (1,1,0,7)} == set(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p,c1'))
+    # Try the same restricting the scan to a single partition instead of
+    # a whole-table scan
+    assert [(0,0,0,1), (0,1,0,3)] == list(cql.execute(f'SELECT p,c1,c2,v FROM {table1} WHERE p=0 GROUP BY p,c1'))
+
+# Adding a LIMIT should be honored.
+# Reproduces #5362 - fewer results than the limit were generated.
+@pytest.mark.xfail(reason="issue #5362")
+def test_group_by_clustering_prefix_with_limit(cql, table1):
+    results = list(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p,c1'))
+    assert len(results) == 4
+    for i in range(1,4):
+        assert results[:i] == list(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p,c1 LIMIT {i}'))
+
+# Adding a PER PARTITION LIMIT should be honored
+# Reproduces #5363 - fewer results than the limit were generated
+@pytest.mark.xfail(reason="issue #5363")
+def test_group_by_clustering_prefix_with_pplimit(cql, table1):
+    # Without per-partition limit we get 4 results, 2 from each partition:
+    assert {(0,0,0,1), (0,1,0,3), (1,0,0,5), (1,1,0,7)} == set(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p,c1'))
+    # With per-partition limit of 2, no change:
+    # But in issue #5363, we got here fewer results.
+    assert {(0,0,0,1), (0,1,0,3), (1,0,0,5), (1,1,0,7)} == set(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p,c1 PER PARTITION LIMIT 2'))
```

Comment:
and here too, this is the exact same query as you are running on line 41... do we need to run it again?